events/schema.py

```diff
@@ -524,6 +527,97 @@ class EventTicketSystemInput(graphene.InputObjectType):
     )
 
 
+class ImportTicketSystemPasswordErrorType(graphene.Scalar):
+    @staticmethod
+    def serialize(errors):
+        if isinstance(errors, dict):
+            if errors.get("__all__", False):
+                errors["non_field_errors"] = errors.pop("__all__")
+            return camelize(errors)
+        raise Exception("`errors` should be dict!")
+
+
+class ImportTicketSystemPasswordsMutation(
+    graphene.relay.ClientIDMutation  # , metaclass=CustomClientIDMutationMeta
+):
+    class Input:
+        event_id = graphene.GlobalID(required=True)
+        passwords = graphene.List(graphene.NonNull(graphene.String), required=True)
+
+    event = graphene.Field(EventNode)
+    passwords = graphene.List(
+        graphene.String, description="A list of imported passwords"
+    )
+    errors = graphene.Field(
+        ImportTicketSystemPasswordErrorType,
+        description="A list of passwords which could not be imported",
+    )
+
+    @classmethod
+    @project_user_required
+    @transaction.atomic
+    def mutate_and_get_payload(cls, root, info, **kwargs):
+        event_id = get_node_id_from_global_id(kwargs.get("event_id"), "EventNode")
+        given_passwords = kwargs.get("passwords")
+
+        if not given_passwords:
+            raise TicketSystemPasswordNothingToImportError(
+                "There is no passwords given to be imported. "
+                "At least 1 password required."
+            )
+
+        try:
+            event = Event.objects.user_can_view(info.context.user).get(pk=event_id)
+        except (Event.DoesNotExist) as e:
+            raise ObjectDoesNotExistError(e)
+
+        # Save passwords one by one instead of bulk, to also collect the errors...
+        (imported_passwords, errors) = cls._create_batch(
+            [
+                TicketSystemPassword(event=event, value=password)
+                for password in given_passwords
+            ]
+        )
+
+        return ImportTicketSystemPasswordsMutation(
+            event=event,
+            passwords=[
+                ticket_system_password.value
+                for ticket_system_password in imported_passwords
+            ],
+            errors=errors,
+        )
+
+    @classmethod
+    def _create_batch(cls, passwords_batch: List[TicketSystemPassword]):
+        imported_passwords: list[TicketSystemPassword] = []
+        passwords_with_errors: list[str] = []
+        for obj in passwords_batch:
+            try:
+                # Use atomic to create a new transaction
+                with transaction.atomic():
```

Comment:
When the loop is processed, if the previous object had an integrity issue for example, the transaction will be locked for the next object. I tried to find a solution to end the transaction when an exception occurs, but couldn't find any. Opening a new transaction was advised. We do that also in other parts of the code.

Reply:
Right, that seems remotely familiar now, thanks for the explanation. I have to say a shorter version of that would be a way more helpful comment there instead of the one just stating the obvious 🙂